### PR TITLE
BIO-48: Support for packed bshd seq dataloader and more advanced fp8 integration for llama3

### PR DESCRIPTION
**Pull request by savitha-eng**
### Description

This PR adds FP8 support enhancements for llama3 training building off https://github.com/NVIDIA/bionemo-framework/pull/1416, including:

1. **Packed BSHD dataloader** - A new `SequencePackingIterableDataset` and `create_bshd_packed_dataloader` that packs sequences by concatenating across boundaries for BSHD format. Unlike THD packing (which tracks boundaries with cu_seqlens), this yields fixed-length samples with no padding, allowing attention to flow across packed sequences. This has been used as a baseline to compare THD to for FP8 experimentation.

2. **BSHD packing toggle in training scripts** - Updated `train_fsdp2.py` and `train_ddp.py` to automatically select the dataloader based on config:
   - `use_sequence_packing=true` + `attn_input_format=bshd` → BSHD packed (cross-boundary attention, no cu_seqlens)
   - `use_sequence_packing=true` + `attn_input_format=thd` → THD packed (respects boundaries via cu_seqlens)
   - `use_sequence_packing=false` → BSHD unpacked (standard windowing)

3. **LM head bf16 for FP8** - Wraps the lm_head forward pass with `fp8_autocast(enabled=False)` to keep it in bf16 for numerical stability during FP8 training. (this is currently also present in https://github.com/NVIDIA/bionemo-framework/pull/1416)

4. **Configurable first/last layer bf16** - Adds `fp8_first_last_bf16` config option that keeps the first and last transformer layers in bf16 while using FP8 for middle layers. This can improve numerical stability during FP8 training.

5. **FP8 tests** - Adds training tests for FP8 in BSHD, THD, and BSHD packed modes, plus a test for the first/last bf16 feature.

#### Usage

# Enable FP8 training with first/last layers in bf16
with initialize_config_dir(config_dir="hydra_config", version_base="1.2"):
    config = compose(
        config_name="L0_sanity",
        overrides=[
            "fp8_config.enabled=true",
            "+dataset.pad_to_multiple_of=16",  # Required for FP8
            "+config_kwargs.fp8_first_last_bf16=true",  # Keep first/last layers in bf16
        ],
    )

# Enable FP8 with BSHD packed dataloader via config
with initialize_config_dir(config_dir="hydra_config", version_base="1.2"):
    config = compose(
        config_name="L0_sanity",
        overrides=[
            "fp8_config.enabled=true",
            "use_sequence_packing=true",
            "config_kwargs.attn_input_format=bshd",  # Triggers BSHD packed dataloader
            "+dataset.pad_to_multiple_of=16",
        ],
    )

# Or use packed BSHD dataloader directly
from dataset import create_bshd_packed_dataloader

dataloader, dataset = create_bshd_packed_dataloader(
    distributed_config=dist_config,
    tokenizer_name_or_path="nvidia/Llama-3.1-8B-Instruct-FP8",
    load_dataset_kwargs={"path": "parquet", "data_files": "data.parquet", "streaming": True},
    micro_batch_size=4,
    max_seq_length=8192,
    pad_to_multiple_of=16,  # For FP8 compatibility
)### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [x] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [x] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 1]
```diff
@@ -275,6 +275,74 @@ def set_epoch(self, epoch: int):
         self.dataset.set_epoch(epoch)
 
 
+@dataclass
+class SequencePackingIterableDataset(torch.utils.data.IterableDataset):
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/src/esm/collator.py:279]: could we just wrap the existing TokenPackingDataset to do this? 

set `max_tokens_per_batch` to be the desired sequence length, set `split_samples`=True; then before you return the sample just concatenate along the sequence dimension

File: bionemo-recipes/recipes/llama3_native_te/dataset.py [hunk 2]
```diff
@@ -334,3 +338,123 @@ def create_cp_dataloader(
         tokenized_dataset = None
 
     return ContextParallelDataLoaderWrapper(train_dataloader, cp_mesh), tokenized_dataset
+
+
+def create_bshd_packed_dataloader(
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/dataset.py:343]: can this be an option in the existing `create_bshd_dataset` function? might be simpler rather than repeating all this. 

I'm wondering though if there's a better way to structure this dataset file where it can be more modular

File: bionemo-recipes/recipes/llama3_native_te/tests/test_train.py [hunk 3]
```diff
@@ -370,6 +370,90 @@ def test_sanity_fsdp2_with_sequence_packing(tmp_path, recipe_path):
...
+def test_train_fsdp2_fp8_bshd(tmp_path, recipe_path):
+    """Test that FSDP2 training works with FP8 enabled in BSHD mode."""
+    with initialize_config_dir(config_dir=str(recipe_path / "hydra_config"), version_base="1.2"):
+        sanity_config = compose(
+            config_name="L0_sanity",
+            overrides=[
+                f"+wandb.dir={tmp_path}",
+                f"checkpoint.ckpt_dir={tmp_path}",
+                "fp8_config.enabled=true",
+                "+dataset.pad_to_multiple_of=16",
```
Comment (pstjohn) [inline bionemo-recipes/recipes/llama3_native_te/tests/test_train.py:381]: well, if you have the fully-packed BSHD dataset, could we just use that here?

**Comment by trvachov** [comment]
Quick question to check my understanding -- "Packed BSHD dataloader" is a feature that is only relevant for llama3/autoregressive models, is that right? I.e. such a dataloader for BERT would not be useful to train an embedding model?

**Comment by pstjohn** [comment]
It mirrors what Megatron-LM does when pre-training large models, which is just concatenating across sequence boundaries without really worrying about document start / ends. Not sure whether BERT would be fine if you trained it like that, but i think the auto-regressive loss makes it slightly less problematic
